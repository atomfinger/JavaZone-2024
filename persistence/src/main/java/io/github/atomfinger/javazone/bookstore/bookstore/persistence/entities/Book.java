package io.github.atomfinger.javazone.bookstore.bookstore.persistence.entities;

import jakarta.persistence.*;

import java.util.Date;

@Entity
@Table(name = "books")
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "book_id")
    private Long bookId;

    @Column(name = "title", nullable = false, length = 255)
    private String title;

    @Column(name = "description")
    private String description;

    @Column(name = "isbn", nullable = false, unique = true, length = 13)
    private String isbn;

    @Column(name = "author_name", nullable = false, length = 255)
    private String authorName;

    @Column(name = "page_numbers")
    private Integer pageNumbers;

    @Column(name = "published_date")
    @Temporal(TemporalType.DATE)
    private Date publishedDate;

    @Column(name = "genre", length = 100)
    private String genre;

    public Long getBookId() {
        return bookId;
    }

    public void setBookId(Long bookId) {
        this.bookId = bookId;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getIsbn() {
        return isbn;
    }

    public void setIsbn(String isbn) {
        this.isbn = isbn;
    }

    public String getAuthorName() {
        return authorName;
    }

    public void setAuthorName(String authorName) {
        this.authorName = authorName;
    }

    public Integer getPageNumbers() {
        return pageNumbers;
    }

    public void setPageNumbers(Integer pageNumbers) {
        this.pageNumbers = pageNumbers;
    }

    public Date getPublishedDate() {
        return publishedDate;
    }

    public void setPublishedDate(Date publishedDate) {
        this.publishedDate = publishedDate;
    }

    public String getGenre() {
        return genre;
    }

    public void setGenre(String genre) {
        this.genre = genre;
    }
}
