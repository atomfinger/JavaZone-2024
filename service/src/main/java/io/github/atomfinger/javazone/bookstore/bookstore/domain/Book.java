package io.github.atomfinger.javazone.bookstore.bookstore.domain;

public class Book {

    private String title, description, author;

    public Book(String title, String description, String author) {
        this.title = title;
        this.description = description;
        this.author = author;
    }

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }

    public String getAuthor() {
        return author;
    }
}
