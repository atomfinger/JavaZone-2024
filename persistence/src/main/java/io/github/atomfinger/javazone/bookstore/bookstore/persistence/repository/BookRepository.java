package io.github.atomfinger.javazone.bookstore.bookstore.persistence.repository;

import io.github.atomfinger.javazone.bookstore.bookstore.persistence.entities.Book;
import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Repository;

@Repository
public interface BookRepository extends CrudRepository<Book, Long> {
}
