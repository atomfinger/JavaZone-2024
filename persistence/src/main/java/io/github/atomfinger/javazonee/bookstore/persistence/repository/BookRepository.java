package io.github.atomfinger.javazonee.bookstore.persistence.repository;

import java.util.Optional;

import org.springframework.data.repository.CrudRepository;
import org.springframework.stereotype.Repository;

import io.github.atomfinger.javazone.bookstore.persistence.entities.Book;

@Repository
public interface BookRepository extends CrudRepository<Book, Long> {

    Optional<Book> findByIsbn(String isbn);
}
