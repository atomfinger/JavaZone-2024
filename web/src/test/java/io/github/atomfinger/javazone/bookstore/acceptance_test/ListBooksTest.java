package io.github.atomfinger.javazone.bookstore.acceptance_test;

import io.github.atomfinger.javazone.bookstore.bookstore.domain.Book;
import org.approvaltests.Approvals;
import org.approvaltests.JsonApprovals;
import org.approvaltests.core.Options;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.util.Assert;

@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class ListBooksTest {

    @LocalServerPort
    private int port;

    @Autowired
    private TestRestTemplate restTemplate;

    @Test
    void greetingShouldReturnDefaultMessage() {
        var result = restTemplate.getForObject("http://localhost:" + port + "/api/books", String.class);
        Approvals.verify(result);
    }
}
