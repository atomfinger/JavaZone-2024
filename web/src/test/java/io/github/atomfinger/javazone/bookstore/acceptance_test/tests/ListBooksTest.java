package io.github.atomfinger.javazone.bookstore.acceptance_test.tests;

import io.github.atomfinger.javazone.bookstore.acceptance_test.AcceptanceTestBase;
import io.github.atomfinger.javazone.bookstore.bookstore.persistence.entities.Book;
import io.github.atomfinger.javazone.bookstore.bookstore.persistence.repository.BookRepository;
import org.approvaltests.JsonApprovals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Calendar;
import java.util.GregorianCalendar;

import static org.mockserver.model.HttpRequest.request;
import static org.mockserver.model.HttpResponse.response;
import static org.springframework.http.HttpHeaders.CONTENT_TYPE;

public class ListBooksTest extends AcceptanceTestBase {

    @Autowired
    BookRepository bookRepository;

    @BeforeEach
    public void setup() {
        var book = new Book();
        book.setTitle("Effective Java");
        book.setDescription("A comprehensive guide to programming in Java.");
        book.setIsbn("9780134685991");
        book.setAuthorName("Joshua Bloch");
        book.setPageNumbers(416);
        book.setPublishedDate(new GregorianCalendar(2018, Calendar.JANUARY, 6).getTime());
        book.setGenre("Programming");
        bookRepository.save(book);
        mockServerClient.when(request().withPath(".*order/.*"))
                .respond(
                        response()
                                .withHeader(CONTENT_TYPE, "application/json; charset=utf-8")
                                .withStatusCode(200)
                                .withBody("""
                                        { "data": [{"isbn": {"code":"9780134685991"}, "order_count": 3}] }
                                        """));
        mockServerClient.when(request().withPath(".*inventory/.*"))
                .respond(
                        response()
                                .withHeader(CONTENT_TYPE, "application/json; charset=utf-8")
                                .withStatusCode(200)
                                .withBody("""
                                        {"inventory": [{"isbn": "9780134685991","count": 5}]}
                                        """)
                );
    }

    @Test
    void given_that_a_book_exists_then_we_should_have_one_listed_out_when_listing_out_books() {
        var result = restTemplate.getForObject(getUrl() + "/books", String.class);
        JsonApprovals.verifyJson(result);
    }
}
