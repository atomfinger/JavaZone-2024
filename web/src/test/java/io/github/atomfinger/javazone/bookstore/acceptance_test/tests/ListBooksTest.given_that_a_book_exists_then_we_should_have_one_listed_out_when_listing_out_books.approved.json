[
  {
    "bookId": 1,
    "title": "Effective Java",
    "description": "A comprehensive guide to programming in Java.",
    "isbn": "9780134685991",
    "authorName": "Joshua Bloch",
    "pageNumbers": 416,
    "publishedDate": "2018-01-06",
    "genre": "Programming",
    "numberOfOrders": 3,
    "isInStock": true
  }
]